File: Top 50 Coding Interview Questions/48. Write code to check if two strings match where one string contains wildcard characters/checkiftwostringsmatchwhereonestringcontainswildcardcharacters.cpp
#include <iostream>
#include <string>
using namespace std;

int main() {
    // Initialize the variables
    string wild, str;

    // Accept the inputs
    cout << "Enter string containing wild characters: ";
    cin >> wild;
    cout << "Enter string to be matched: ";
    cin >> str;

    int m = wild.length();
    int n = str.length();
    
    // Create a 2D array for dynamic programming
    bool check[m + 1][n + 1] = {false};

    // Initialize the DP table
    check[0][0] = true;

    // Handle the case where wild characters start with '*'
    for (int i = 1; i <= m; i++) {
        if (wild[i - 1] == '*') {
            check[i][0] = check[i - 1][0];
        } else {
            check[i][0] = false;
        }
    }

    // Fill the DP table
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (wild[i - 1] == str[j - 1] || wild[i - 1] == '?') {
                check[i][j] = check[i - 1][j - 1];
            } else if (wild[i - 1] == '*') {
                check[i][j] = check[i - 1][j] || check[i][j - 1];
            } else {
                check[i][j] = false;
            }
        }
    }

    // Printing the result
    if (check[m][n]) {
        cout << "TRUE";
    } else {
        cout << "FALSE";
    }

    return 0;
}

File: Top 50 Coding Interview Questions/48. Write code to check if two strings match where one string contains wildcard characters/checkiftwostringsmatchwhereonestringcontainswildcardcharacters.py
def solve(a, b):
    """
    Recursively checks if the string 'b' matches the pattern 'a',
    where 'a' can contain wildcards '*' and '?'.

    Args:
    a (str): Pattern string containing wildcards ('*', '?').
    b (str): Target string to match.

    Returns:
    bool: True if the pattern matches the string, False otherwise.
    """
    n, m = len(a), len(b)

    # If both strings are empty, they match.
    if n == 0 and m == 0:
        return True

    # If pattern contains '*' but target is empty, check the rest of the pattern.
    if n > 1 and a[0] == '*' and m == 0:
        return solve(a[1:], b)

    # If the first character matches or if '?' is used as a wildcard.
    if (n > 0 and m > 0 and a[0] == b[0]) or (n > 0 and a[0] == '?'):
        return solve(a[1:], b[1:])

    # If '*' is encountered, it can match zero or more characters.
    if n > 0 and a[0] == '*':
        return solve(a[1:], b) or solve(a, b[1:])

    # If none of the above cases match, the pattern and string don't match.
    return False

# Take user input for both strings.
str1 = input("Enter the first string (with wild characters): ")
str2 = input("Enter the second string (without wild characters): ")

# Print the input strings.
print("First string with wild characters:", str1)
print("Second string without wild characters:", str2)

# Call the solve function and print the result.
if solve(str1, str2):
    print("The strings match!")
else:
    print("The strings do not match.")
